domain/keymanager/modelmigration/import.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package modelmigration
+
+import (
+	"context"
+	"strings"
+
+	"github.com/juju/description/v8"
+
+	"github.com/juju/juju/core/modelmigration"
+	"github.com/juju/juju/core/user"
+	accesserrors "github.com/juju/juju/domain/access/errors"
+	accessservice "github.com/juju/juju/domain/access/service"
+	accessstate "github.com/juju/juju/domain/access/state"
+	"github.com/juju/juju/domain/controller"
+	"github.com/juju/juju/domain/keymanager/service"
+	"github.com/juju/juju/domain/keymanager/state"
+	"github.com/juju/juju/internal/errors"
+	"github.com/juju/juju/internal/ssh"
+)
+
+const (
+	// modelConfigKeyAuthorizedKeys is the old model config key that was used
+	// to describe authorized keys for a model in model config. This key has
+	// been removed since and now resides here for backwards compatibility with
+	// 3.x controllers.
+	modelConfigKeyAuthorizeKeys = "authorized-keys"
+)
+
+// Coordinator is the interface that is used to add operations to a migration.
+type Coordinator interface {
+	// Add adds the given operation to the migration.
+	Add(modelmigration.Operation)
+}
+
+// ImportService represents the service methods needed for importing the
+// authorized keys of a model during migration.
+type ImportService interface {
+	// AddPublicKeysForUser is responsible for adding public keys for a user to a
+	// model. The following errors can be expected:
+	// - [errors.NotValid] when the user id is not valid
+	// - [github.com/juju/juju/domain/access/errors.UserNotFound] when the user does
+	// not exist.
+	// - [keyserrors.InvalidPublicKey] when a public key fails validation.
+	// - [keyserrors.ReservedCommentViolation] when a key being added contains a
+	// comment string that is reserved.
+	// - [keyserrors.PublicKeyAlreadyExists] when a public key being added
+	// for a user already exists.
+	AddPublicKeysForUser(context.Context, user.UUID, ...string) error
+}
+
+// UserService represents the service methods needed for finding users when
+// importing ssh public keys for a model.
+type UserService interface {
+	// GetUserByName will find and return the user associated with name. If
+	// there is no user for the user name then an error that satisfies
+	// [github.com/juju/juju/domain/access/errors].NotFound will be returned.
+	// If supplied with an invalid user name then an error that satisfies
+	// [github.com/juju/juju/domain/access/errors].UserNameNotValid will be
+	// returned.
+	GetUserByName(context.Context, user.Name) (user.User, error)
+
+	// GetUsernamesForIds is responsible for returning all of the [user.Name]
+	// for each of the [user.UUID] passed to this function. Should one of the
+	// user id's not exist then the whole call fails with a
```

Comment:
apostrophe.
 Ideally this comment would be the same as that on the service method itself.